source/unified-test-format/unified-test-format.rst

```diff
@@ -0,0 +1,2515 @@
+
+
+test
+~~~~
+
+Test case consisting of a sequence of operations to be executed.
+
+The structure of each document is as follows:
+
+- ``description``: Required string. The name of the test.
+
+  This SHOULD describe the purpose of this test (e.g. "insertOne is retried").
+
+.. _test_runOn:
+
+- ``runOn``: Optional array of documents. List of server version and/or topology
+  requirements for which the tests in this file can be run. If specified, these
+  requirements are evaluated independently and in addition to any top-level
+  `runOn`_ requirements. If no requirements in this array are met, the test
+  runner MUST skip this test.
+
+  These requirements SHOULD be more restrictive than those specified in the
+  top-level `runOn`_ requirements (if any). They SHOULD NOT be more permissive.
+  This is advised because both sets of requirements MUST be satisified in order
+  for a test to be executed and more permissive requirements at the test-level
+  could be taken out of context on their own.
+
+  If set, the array should contain at least one document. The structure of each
+  document is defined in `runOnRequirement`_.
+
+.. _test_skipReason:
+
+- ``skipReason``: Optional string. If set, the test will be skipped. The string
+  SHOULD explain the reason for skipping the test (e.g. JIRA ticket).
+
+.. _test_operations:
+
+- ``operations``: Required array of documents. List of operations to be executed
+  for the test case.
+
+  The array should contain at least one document. The structure of each
+  document is defined in `operation`_.
+
+.. _test_expectedEvents:
+
+- ``expectedEvents``: Optional array of documents. Each document will specify a
+  client entity and a list of events that are expected to be observed (in that
+  order) on that client while executing `operations <test_operations_>`_.
+
+  If a driver only supports configuring event listeners globally (for all
+  clients), the test runner SHOULD associate each observed event with a client
+  in order to perform these assertions.
+
+  Test files SHOULD NOT expect events from multiple specs (e.g. command
+  monitoring *and* SDAM events) for a single client. See
+  `Mixing event types in observeEvents and expectedEvents`_ for more
+  information.
+
+  The array should contain at least one document. The structure of each document
+  is as follows:
+
+  - ``client``: Required string. Client entity on which the events are expected
+    to be observed. See `commonOptions_client`_.
+
+  - ``events``: Required array of documents. List of events, which are expected
+    to be observed (in this order) on the corresponding client while executing
+    `operations`_. If the array is empty, the test runner MUST assert that no
+    events were observed on the client (excluding ignored events).
+
+    The structure of each document is defined in `expectedEvent`_.
+
+.. _test_outcome:
+
+- ``outcome``: Optional array of documents. Each document will specify expected
```

Comment:
Similar to my input on `initialData`, here `outcome`, could be an array of follow-up operations with matchers rather then a custom implementation to check the collection, this would all the test to check more then just collection data, and be able to run any method, and run expectations against them.

Reply:
Interesting idea. Note that `initialData` and `outcome` never use the entities under test and are intended to use a separate MongoClient, so changing this would require most tests to create separate entities just for these pre- and post-test routines.

As for outcome assertions, there's nothing stopping a test from running `find` at the end of `operations` to utilize matcher logic. Likewise, tests can do arbitrary CRUD ops to populate or alter initial data. This assumes that the test doesn't care about using an entity under test and is capable of ignoring any extra command monitoring events these operations might produce (`ignoreCommandMonitoringEvents` can probably help, assuming the test observes events at all). I am expecting to use both of those techniques for GridFS tests.

Reply:
Thanks for your detailed response here, I understand the extra setup configuration would be an extra step. The philosophy would be more configuration over the conventions of `outcome` and `initialData`. But I see that there's something to having these keys, and the functionality that makes the specs easier to write.